File: src/longest_substring.js
const lengthOfLongestSubstring = (s) => {

    const cache = new Map();

    let start = 0;
    let max = 0;

    for (let end = 0; end < s.length; end ++) {
      if (cache.has(s[end]) && cache.get(s[end]) >= start) {

        max = Math.max(res, end - start);
        start = cache.get( s[end] ) + 1;
      }

      cache.set(s[end], end);
    }

    return Math.max(max, s.length - start);
  };

console.log(lengthOfLongestSubstring("au"));

File: src/longest_substring.rs
use std::{collections::HashMap, cmp};



pub struct Solution;

impl Solution {
    pub fn length_of_longest_substring(s: String) -> i32 {

        let mut cache: HashMap< char, usize > = HashMap::new();
        let mut start: usize = 0;
        let mut  max = 0;

        for (end, letter) in s.chars().enumerate() {
            if cache.contains_key(&letter) && *cache.get(&letter).unwrap() >= start {
                max = cmp::max(max,  end - start );
                start = *cache.get(&letter).unwrap() + 1;
            }

            cache.insert(letter, end);
        }

        return cmp::max(max,  s.len() - start ) as i32;
        
    }
}

#[test]
fn should_word() {

}
